First steps to layerize canvas

emilio opened the pull request:
I've done a bit of job to get this done. Right now readback is still used, but we have a `LayerId` -> `CanvasRenderer` map on the paint task, that we can use to get rid of that.

I'd want review, to see if this is a good approach (I know it's not the initial `CanvasId` -> renderer approach, but it's pretty similar, since a canvas involves a `PaintLayer`).

I had to do a bit of refactoring to avoid cyclic dependencies between canvas and gfx. I'd want you to review them too.

It's mergeable and doesn't break any tests :P

Some of my main concerns:
- Does the canvas render really need to be behind an `Arc<Mutex<T>>`?
- I can't clone a `NativeSurface` right now (that's why the `SendNativeSurface()` msg is unimplemented in the WebGL task). It should be easy to add that to rust-layers, supposing the caller is responsible to mark it as non-leaking, any reason to not do it?

cc @jdm @pcwalton

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6083)

<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -184,6 +190,16 @@ impl WebGLPaintTask {
+        //                             .borrow_bound_surface().unwrap();
+        // let surface = (*surface).clone();
+        // surface.mark_wont_leak();
+        // chan.send(surface).unwrap()
```
I'd rather not have a bunch of commented code around—it tends to bitrot quickly. Could you remove this?

pcwalton commented:
I'm confused as to what the tests are testing.

ecoal95 commented:
Yep, I'm sorry. My first try was to generate the layer before calling create_stacking_context, and use an `expect()` from there if its a canvas fragment, but that required a lot of checks depending of every situation that could generate a layer, so I opted by the potential_layer_id approach. I forgot to remove them. I'll do it right now,  as well as the commented code :P

pcwalton commented on components/layout/display_list_builder.rs:
```diff
@@ -1088,8 +1088,11 @@ impl FragmentDisplayListBuilding for Fragment {
+                               potential_layer_id: LayerId,
+                               layer: Option<PaintLayer>)
                                -> Arc<StackingContext> {
+        debug!("create_stacking_context");
```
Seems like not a very useful debug message; could you remove this?

pcwalton commented on components/layout/display_list_builder.rs:
```diff
@@ -1120,6 +1123,24 @@ impl FragmentDisplayListBuilding for Fragment {
+        // If it's a canvas we must propagate the layer and the renderer to the paint
+        // task
+        match self.specific {
+            SpecificFragmentInfo::Canvas(ref canvas_fragment_info) => {
```
nit: use `if let`

pcwalton commented on components/layout/display_list_builder.rs:
```diff
@@ -1120,6 +1123,24 @@ impl FragmentDisplayListBuilding for Fragment {
 
+        // Ensure every canvas has a layer
+        let layer = match self.specific {
+            SpecificFragmentInfo::Canvas(_) => layer.or(Some(PaintLayer::new(potential_layer_id, color::transparent(), ScrollPolicy::Scrollable))),
```
nit: I think this line is over 100 characters

pcwalton commented on components/layout/display_list_builder.rs:
```diff
@@ -1088,8 +1088,11 @@ impl FragmentDisplayListBuilding for Fragment {
-                               layer: Option<Arc<PaintLayer>>)
+                               layout_context: &LayoutContext,
+                               potential_layer_id: LayerId,
+                               layer: Option<PaintLayer>)
```
Looks like `potential_layer_id` is never used unless `layer` is `None`. Could you refactor it to something like this:

    enum LayerForStackingContext {
        Existing(PaintLayer),
        New(LayerId),
    }

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1976,6 +1974,11 @@ impl Fragment {
             return true
         }
+        // Canvas always layerizes, as an special case
+        match self.specific {
```
nit: `if let`

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1976,6 +1974,11 @@ impl Fragment {
         if self.style().get_effects().transform.is_some() {
             return true
         }
+        // Canvas always layerizes, as an special case
```
Could you add something like "FIXME(pcwalton): Don't unconditionally form stacking contexts for each canvas." here? @rocallahan wants us to fix this at some point and I agree with him. :) I'd be happy to elaborate more on the long-term plans if you're interested.

pcwalton commented:
The overall approach looks great; just a few mostly-stylistic nits. Awesome work!

ecoal95 commented:
I addressed all of your comments. I also added a debug message when a renderer is received, to keep track of them :P

And of course I'm interested in hearing about the long-term plans about this :)

pcwalton commented:
@ecoal95 I filed https://github.com/servo/servo/issues/6142.

pcwalton commented:
Looks good; squash away!

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit b4d6721 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton b4d6721482c5da9aab03ccc41476def50cf35ed3 -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6053) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:lock: Merge conflict

emilio committed:
Layerize canvas

Note that this keeps using readback right now, `NativeSurface` painting
will be implemented soon.

Also see https://github.com/servo/servo/issues/6142

jdm commented:
@bors-servo: r=pcwalton

bors-servo commented:
:pushpin: Commit 1562994 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 1562994184f3e83dda53f1b7e7f5246228fbbb56 -->

bors-servo commented:
:hourglass: Testing commit 1562994 with merge 04392bc...

bors-servo commented:
:broken_heart: Test failed - [mac1](http://build.servo.org/builders/mac1/builds/2016)

jdm commented:
Lots of failures:
```
/_mozilla/mozilla/enabled_disabled_selectors.html
-------------------------------------------------
TIMEOUT [Parent]
/dom/nodes/Element-matches.html
-------------------------------
TIMEOUT [Parent]
/dom/nodes/ParentNode-querySelector-All.html
--------------------------------------------
TIMEOUT [Parent]
/html/infrastructure/common-dom-interfaces/collections/htmlformcontrolscollection.html
--------------------------------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/forms/constraints/form-validation-validity-tooLong.html
-----------------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/forms/constraints/form-validation-validity-tooShort.html
------------------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/forms/the-form-element/form-nameditem.html
----------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/forms/the-textarea-element/textarea-type.html
-------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/grouping-content/the-li-element/grouping-li.html
----------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol.html
----------------------------------------------------------------
TIMEOUT [Parent]
/html/rendering/bindings/the-textarea-element-0/cols-default.html
-----------------------------------------------------------------
TIMEOUT [Parent]
/html/rendering/bindings/the-textarea-element-0/cols-zero.html
--------------------------------------------------------------
TIMEOUT [Parent]
/html/rendering/bindings/the-textarea-element-0/rows-default.html
-----------------------------------------------------------------
TIMEOUT [Parent]
/html/rendering/bindings/the-textarea-element-0/rows-zero.html
--------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol-start-reftest-001.html
----------------------------------------------------------------------------------
CRASH expected FAIL [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol-start-reftest-002.html
----------------------------------------------------------------------------------
CRASH expected FAIL [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol-type-reftest-001.html
---------------------------------------------------------------------------------
CRASH [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol-type-reftest-002.html
---------------------------------------------------------------------------------
CRASH expected FAIL [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol-type-reftest-003.html
---------------------------------------------------------------------------------
CRASH expected FAIL [Parent]
/html/semantics/grouping-content/the-li-element/grouping-li-reftest-002.html
----------------------------------------------------------------------------
TIMEOUT [Parent]
/html/semantics/grouping-content/the-ol-element/grouping-ol-rev-reftest-001.html
--------------------------------------------------------------------------------
TIMEOUT expected FAIL [Parent]
```
eg.
```
0:19.93 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) Full command: /Users/servo/buildbot/slave/mac1/build/components/servo/target/debug/servo --cpu --hard-fail -z http://localhost:8000/_mozilla/mozilla/enabled_disabled_selectors.html
(pid:3413) "thread 'LayoutWorker worker 6/6' panicked at 'index out of bounds: the len is 0 but the index is 0', thread 'LayoutWorker worker 1/6/Users/larsberg/rust/src/libcollections/vec.rs:1349"
 0:19.93 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "thread '' panicked at 'stack backtrace:"
 0:19.93 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "LayoutWorker worker 2/6index out of bounds: the len is 0 but the index is 0', thread 'LayoutWorker worker 5/6' panicked at 'index out of bounds: the len is 0 but the index is 0', /Users/larsberg/rust/src/libcollections/vec.rs/Users/larsberg/rust/src/libcollections/vec.rs::13491349"
 0:19.93 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "' panicked at 'index out of bounds: the len is 0 but the index is 0', /Users/larsberg/rust/src/libcollections/vec.rs"
 0:19.93 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) ":1349"
 0:19.93 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "thread 'LayoutWorker worker 4/6' panicked at 'index out of bounds: the len is 0 but the index is 0', /Users/larsberg/rust/src/libcollections/vec.rs:1349"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   1:        0x110b2a75f - sys::backtrace::write::h6990067d88907b48zVr"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   2:        0x110b2e134 - panicking::on_panic::he976cdbadffd9e1bEVv"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   3:        0x110b19c15 - rt::unwind::begin_unwind_inner::ha519446d83876aa8nDv"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   4:        0x110b1a44c - rt::unwind::begin_unwind_fmt::h15c2d4adf8e8ecdatCv"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   5:        0x110b2d9ec - rust_begin_unwind"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   6:        0x110b52e95 - panicking::panic_fmt::haa553fbfab6eff2cNKy"
 0:19.95 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   7:        0x110b55078 - panicking::panic_bounds_check::ha04a96a9aae86ca6TJy"
 0:19.96 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   8:        0x10ee70426 - vec::Vec<T>.Index<usize>::index::h4395348926812132377"
 0:19.96 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   9:        0x10ee6fb57 - display_list_builder::InlineFlow.InlineFlowDisplayListBuilding::build_display_list_for_inline::he6d046e584578146n3f"
 0:19.96 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  10:        0x10ed9a6f1 - inline::InlineFlow.Flow::build_display_list::h045055fdd4346a65Lbp"
 0:19.96 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  11:        0x10ef777a2 - traversal::BuildDisplayList<'a>.PostorderFlowTraversal::process::h44f52c748227c02f63u"
 0:19.96 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  12:        0x10ef774f8 - parallel::ParallelPostorderFlowTraversal::run_parallel::h17925963689472074348"
 0:19.96 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  13:        0x10ef74dcd - parallel::build_display_list::h77d2b372ca90091eUxq"
 0:19.97 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  14:        0x10ef74b6f - parallel::ComputeAbsolutePositions<'a>.ParallelPreorderFlowTraversal::run_parallel::hc3b5891b703a2d59ztq"
 0:19.97 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  15:        0x10ef74d1d - parallel::compute_absolute_positions::hac5998288a1939ebgxq"
 0:19.97 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  16:        0x10eefb590 - workqueue::WorkerThread<QueueData, WorkData>::start::h15607962942249882032"
 0:19.97 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  17:        0x10eefadb3 - workqueue::WorkQueue<QueueData, WorkData>::new::closure.41530"
 0:19.97 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  18:        0x10eefac06 - task::spawn_named::closure.41524"
 0:19.97 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  19:        0x10eefab0e - boxed::F.FnBox<A>::call_box::h16865387641012896107"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  20:        0x10eeceabb - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h6085128213034823021"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  21:        0x10eece11b - thread::Builder::spawn_inner::closure.40742"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  22:        0x10eece09e - rt::unwind::try::try_fn::__rust_abi::h17004706174750548169"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  23:        0x10eece039 - rt::unwind::try::try_fn::h17004706174750548169"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  24:        0x110b2fe28 - rust_try_inner"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  25:        0x110b2fe15 - rust_try"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  26:        0x10eecd4fa - rt::unwind::try::h10063819792191526983"
 0:19.98 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  27:        0x10eecd2a4 - thread::Builder::spawn_inner::closure.40673"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  28:        0x10eecedfd - boxed::F.FnBox<A>::call_box::h4227202771269554373"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  29:        0x110b2cabd - sys::thread::Thread::new::thread_start::h5fcf8fe12ec3c274mYu"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  30:     0x7fff8fbd7771 - _pthread_start"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "stack backtrace:"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   1:        0x110b2a75f - sys::backtrace::write::h6990067d88907b48zVr"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   2:        0x110b2e134 - panicking::on_panic::he976cdbadffd9e1bEVv"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   3:        0x110b19c15 - rt::unwind::begin_unwind_inner::ha519446d83876aa8nDv"
 0:19.99 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   4:        0x110b1a44c - rt::unwind::begin_unwind_fmt::h15c2d4adf8e8ecdatCv"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   5:        0x110b2d9ec - rust_begin_unwind"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   6:        0x110b52e95 - panicking::panic_fmt::haa553fbfab6eff2cNKy"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   7:        0x110b55078 - panicking::panic_bounds_check::ha04a96a9aae86ca6TJy"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   8:        0x10ee70426 - vec::Vec<T>.Index<usize>::index::h4395348926812132377"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   9:        0x10ee6fb57 - display_list_builder::InlineFlow.InlineFlowDisplayListBuilding::build_display_list_for_inline::he6d046e584578146n3f"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  10:        0x10ed9a6f1 - inline::InlineFlow.Flow::build_display_list::h045055fdd4346a65Lbp"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  11:        0x10ef777a2 - traversal::BuildDisplayList<'a>.PostorderFlowTraversal::process::h44f52c748227c02f63u"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  12:        0x10ef774f8 - parallel::ParallelPostorderFlowTraversal::run_parallel::h17925963689472074348"
 0:20.00 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  13:        0x10ef74dcd - parallel::build_display_list::h77d2b372ca90091eUxq"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  14:        0x10ef74b6f - parallel::ComputeAbsolutePositions<'a>.ParallelPreorderFlowTraversal::run_parallel::hc3b5891b703a2d59ztq"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  15:        0x10ef74d1d - parallel::compute_absolute_positions::hac5998288a1939ebgxq"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  16:        0x10eefb590 - workqueue::WorkerThread<QueueData, WorkData>::start::h15607962942249882032"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  17:        0x10eefadb3 - workqueue::WorkQueue<QueueData, WorkData>::new::closure.41530"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  18:        0x10eefac06 - task::spawn_named::closure.41524"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  19:        0x10eefab0e - boxed::F.FnBox<A>::call_box::h16865387641012896107"
 0:20.01 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  20:        0x10eeceabb - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h6085128213034823021"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  21:        0x10eece11b - thread::Builder::spawn_inner::closure.40742"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  22:        0x10eece09e - rt::unwind::try::try_fn::__rust_abi::h17004706174750548169"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  23:        0x10eece039 - rt::unwind::try::try_fn::h17004706174750548169"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  24:        0x110b2fe28 - rust_try_inner"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  25:        0x110b2fe15 - rust_try"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  26:        0x10eecd4fa - rt::unwind::try::h10063819792191526983"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  27:        0x10eecd2a4 - thread::Builder::spawn_inner::closure.40673"
 0:20.02 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  28:        0x10eecedfd - boxed::F.FnBox<A>::call_box::h4227202771269554373"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  29:        0x110b2cabd - sys::thread::Thread::new::thread_start::h5fcf8fe12ec3c274mYu"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  30:     0x7fff8fbd7771 - _pthread_start"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "stack backtrace:"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   1:        0x110b2a75f - sys::backtrace::write::h6990067d88907b48zVr"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   2:        0x110b2e134 - panicking::on_panic::he976cdbadffd9e1bEVv"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   3:        0x110b19c15 - rt::unwind::begin_unwind_inner::ha519446d83876aa8nDv"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   4:        0x110b1a44c - rt::unwind::begin_unwind_fmt::h15c2d4adf8e8ecdatCv"
 0:20.03 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   5:        0x110b2d9ec - rust_begin_unwind"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   6:        0x110b52e95 - panicking::panic_fmt::haa553fbfab6eff2cNKy"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   7:        0x110b55078 - panicking::panic_bounds_check::ha04a96a9aae86ca6TJy"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   8:        0x10ee70426 - vec::Vec<T>.Index<usize>::index::h4395348926812132377"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   9:        0x10ee6fb57 - display_list_builder::InlineFlow.InlineFlowDisplayListBuilding::build_display_list_for_inline::he6d046e584578146n3f"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  10:        0x10ed9a6f1 - inline::InlineFlow.Flow::build_display_list::h045055fdd4346a65Lbp"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  11:        0x10ef777a2 - traversal::BuildDisplayList<'a>.PostorderFlowTraversal::process::h44f52c748227c02f63u"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  12:        0x10ef774f8 - parallel::ParallelPostorderFlowTraversal::run_parallel::h17925963689472074348"
 0:20.04 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  13:        0x10ef74dcd - parallel::build_display_list::h77d2b372ca90091eUxq"
 0:20.05 TEST_END: Thread-TestrunnerManager-1 Harness OK. Subtests passed 1/1. Unexpected 0
 0:20.05 TEST_START: Thread-TestrunnerManager-1 /_mozilla/mozilla/event_handler_syntax_error.html
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  14:        0x10ef74b6f - parallel::ComputeAbsolutePositions<'a>.ParallelPreorderFlowTraversal::run_parallel::hc3b5891b703a2d59ztq"
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  15:        0x10ef74d1d - parallel::compute_absolute_positions::hac5998288a1939ebgxq"
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  16:        0x10eefb590 - workqueue::WorkerThread<QueueData, WorkData>::start::h15607962942249882032"
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  17:        0x10eefadb3 - workqueue::WorkQueue<QueueData, WorkData>::new::closure.41530"
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  18:        0x10eefac06 - task::spawn_named::closure.41524"
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  19:        0x10eefab0e - boxed::F.FnBox<A>::call_box::h16865387641012896107"
 0:20.05 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  20:        0x10eeceabb - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h6085128213034823021"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  21:        0x10eece11b - thread::Builder::spawn_inner::closure.40742"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  22:        0x10eece09e - rt::unwind::try::try_fn::__rust_abi::h17004706174750548169"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  23:        0x10eece039 - rt::unwind::try::try_fn::h17004706174750548169"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  24:        0x110b2fe28 - rust_try_inner"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  25:        0x110b2fe15 - rust_try"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  26:        0x10eecd4fa - rt::unwind::try::h10063819792191526983"
 0:20.06 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  27:        0x10eecd2a4 - thread::Builder::spawn_inner::closure.40673"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  28:        0x10eecedfd - boxed::F.FnBox<A>::call_box::h4227202771269554373"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  29:        0x110b2cabd - sys::thread::Thread::new::thread_start::h5fcf8fe12ec3c274mYu"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  30:     0x7fff8fbd7771 - _pthread_start"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "stack backtrace:"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   1:        0x110b2a75f - sys::backtrace::write::h6990067d88907b48zVr"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   2:        0x110b2e134 - panicking::on_panic::he976cdbadffd9e1bEVv"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   3:        0x110b19c15 - rt::unwind::begin_unwind_inner::ha519446d83876aa8nDv"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   4:        0x110b1a44c - rt::unwind::begin_unwind_fmt::h15c2d4adf8e8ecdatCv"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   5:        0x110b2d9ec - rust_begin_unwind"
 0:20.07 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   6:        0x110b52e95 - panicking::panic_fmt::haa553fbfab6eff2cNKy"
 0:20.08 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   7:        0x110b55078 - panicking::panic_bounds_check::ha04a96a9aae86ca6TJy"
 0:20.08 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   8:        0x10ee70426 - vec::Vec<T>.Index<usize>::index::h4395348926812132377"
 0:20.08 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   9:        0x10ee6fb57 - display_list_builder::InlineFlow.InlineFlowDisplayListBuilding::build_display_list_for_inline::he6d046e584578146n3f"
 0:20.08 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  10:        0x10ed9a6f1 - inline::InlineFlow.Flow::build_display_list::h045055fdd4346a65Lbp"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  11:        0x10ef777a2 - traversal::BuildDisplayList<'a>.PostorderFlowTraversal::process::h44f52c748227c02f63u"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  12:        0x10ef774f8 - parallel::ParallelPostorderFlowTraversal::run_parallel::h17925963689472074348"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  13:        0x10ef74dcd - parallel::build_display_list::h77d2b372ca90091eUxq"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  14:        0x10ef74b6f - parallel::ComputeAbsolutePositions<'a>.ParallelPreorderFlowTraversal::run_parallel::hc3b5891b703a2d59ztq"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  15:        0x10ef74d1d - parallel::compute_absolute_positions::hac5998288a1939ebgxq"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  16:        0x10eefb590 - workqueue::WorkerThread<QueueData, WorkData>::start::h15607962942249882032"
 0:20.09 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  17:        0x10eefadb3 - workqueue::WorkQueue<QueueData, WorkData>::new::closure.41530"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  18:        0x10eefac06 - task::spawn_named::closure.41524"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  19:        0x10eefab0e - boxed::F.FnBox<A>::call_box::h16865387641012896107"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  20:        0x10eeceabb - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h6085128213034823021"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  21:        0x10eece11b - thread::Builder::spawn_inner::closure.40742"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  22:        0x10eece09e - rt::unwind::try::try_fn::__rust_abi::h17004706174750548169"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  23:        0x10eece039 - rt::unwind::try::try_fn::h17004706174750548169"
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  24:        0x110b2fe28 - rust_try_inner"
 0:20.10 TEST_END: Thread-TestrunnerManager-3 Harness OK. Subtests passed 1/1. Unexpected 0
 0:20.10 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  25:        0x110b2fe15 - rust_try"
 0:20.10 TEST_START: Thread-TestrunnerManager-3 /_mozilla/mozilla/event_listener.html
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  26:        0x10eecd4fa - rt::unwind::try::h10063819792191526983"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  27:        0x10eecd2a4 - thread::Builder::spawn_inner::closure.40673"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  28:        0x10eecedfd - boxed::F.FnBox<A>::call_box::h4227202771269554373"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  29:        0x110b2cabd - sys::thread::Thread::new::thread_start::h5fcf8fe12ec3c274mYu"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  30:     0x7fff8fbd7771 - _pthread_start"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "stack backtrace:"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   1:        0x110b2a75f - sys::backtrace::write::h6990067d88907b48zVr"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   2:        0x110b2e134 - panicking::on_panic::he976cdbadffd9e1bEVv"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   3:        0x110b19c15 - rt::unwind::begin_unwind_inner::ha519446d83876aa8nDv"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   4:        0x110b1a44c - rt::unwind::begin_unwind_fmt::h15c2d4adf8e8ecdatCv"
 0:20.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   5:        0x110b2d9ec - rust_begin_unwind"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   6:        0x110b52e95 - panicking::panic_fmt::haa553fbfab6eff2cNKy"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   7:        0x110b55078 - panicking::panic_bounds_check::ha04a96a9aae86ca6TJy"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   8:        0x10ee70426 - vec::Vec<T>.Index<usize>::index::h4395348926812132377"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "   9:        0x10ee6fb57 - display_list_builder::InlineFlow.InlineFlowDisplayListBuilding::build_display_list_for_inline::he6d046e584578146n3f"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  10:        0x10ed9a6f1 - inline::InlineFlow.Flow::build_display_list::h045055fdd4346a65Lbp"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  11:        0x10ef777a2 - traversal::BuildDisplayList<'a>.PostorderFlowTraversal::process::h44f52c748227c02f63u"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  12:        0x10ef774f8 - parallel::ParallelPostorderFlowTraversal::run_parallel::h17925963689472074348"
 0:20.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  13:        0x10ef74dcd - parallel::build_display_list::h77d2b372ca90091eUxq"
 0:20.13 TEST_END: Thread-TestrunnerManager-4 Harness OK. Subtests passed 1/1. Unexpected 0
 0:20.13 TEST_START: Thread-TestrunnerManager-4 /_mozilla/mozilla/focus_blur.html
 0:20.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  14:        0x10ef74b6f - parallel::ComputeAbsolutePositions<'a>.ParallelPreorderFlowTraversal::run_parallel::hc3b5891b703a2d59ztq"
 0:20.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  15:        0x10ef74d1d - parallel::compute_absolute_positions::hac5998288a1939ebgxq"
 0:20.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  16:        0x10eefb590 - workqueue::WorkerThread<QueueData, WorkData>::start::h15607962942249882032"
 0:20.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  17:        0x10eefadb3 - workqueue::WorkQueue<QueueData, WorkData>::new::closure.41530"
 0:20.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  18:        0x10eefac06 - task::spawn_named::closure.41524"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  19:        0x10eefab0e - boxed::F.FnBox<A>::call_box::h16865387641012896107"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  20:        0x10eeceabb - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h6085128213034823021"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  21:        0x10eece11b - thread::Builder::spawn_inner::closure.40742"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  22:        0x10eece09e - rt::unwind::try::try_fn::__rust_abi::h17004706174750548169"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  23:        0x10eece039 - rt::unwind::try::try_fn::h17004706174750548169"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  24:        0x110b2fe28 - rust_try_inner"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  25:        0x110b2fe15 - rust_try"
 0:20.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  26:        0x10eecd4fa - rt::unwind::try::h10063819792191526983"
 0:20.15 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  27:        0x10eecd2a4 - thread::Builder::spawn_inner::closure.40673"
 0:20.15 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  28:        0x10eecedfd - boxed::F.FnBox<A>::call_box::h4227202771269554373"
 0:20.15 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  29:        0x110b2cabd - sys::thread::Thread::new::thread_start::h5fcf8fe12ec3c274mYu"
 0:20.15 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:3413) "  30:     0x7fff8fbd7771 - _pthread_start"
```

ecoal95 commented:
My fault, didn't consider inline flows without fragments. It should be all working now, I passed the wpt test locally without problems.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 3350522 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 3350522306ee087f892108ee82aeb9d58ee2363d -->

bors-servo commented:
:hourglass: Testing commit 3350522 with merge ad53e95...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1825), [gonk](http://build.servo.org/builders/gonk/builds/982), [linux1](http://build.servo.org/builders/linux1/builds/1815), [linux2](http://build.servo.org/builders/linux2/builds/1816), [mac1](http://build.servo.org/builders/mac1/builds/2023), [mac2](http://build.servo.org/builders/mac2/builds/2029)

dmarcos commented:
@ecoal95 @pcwalton Can you describe the remaining work to get proper layerization of canvas? I can jump and help but I need more clarity on the next steps. Thanks!
